cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java

```diff
@@ -103,15 +104,40 @@ private double balancePercentageWithMargin(OptimizationOptions optimizationOptio
     return (balancePercentage - 1) * BALANCE_MARGIN;
   }
 
+  /**
+   * Ensure that the given balance limit falls into min/max limits determined by min/max gaps for topic replica balance.
+   * If the computed balance limit is out of these gap-based limits, use the relevant max/min gap-based balance limit.
+   *
+   * @see com.linkedin.kafka.cruisecontrol.config.constants.AnalyzerConfig#TOPIC_REPLICA_COUNT_BALANCE_MIN_GAP_DOC
+   * @see com.linkedin.kafka.cruisecontrol.config.constants.AnalyzerConfig#TOPIC_REPLICA_COUNT_BALANCE_MAX_GAP_DOC
+   *
+   * @param computedLimit Computed balance upper or lower limit
+   * @param average Average topic replicas on broker.
+   * @return A balance limit that falls into [minGap, maxGap] for topic replica balance.
+   */
+  private int gapBasedBalanceLimit(int computedLimit, double average, boolean isLowerLimit) {
+    int minLimit;
+    int maxLimit;
+    if (isLowerLimit) {
+      minLimit = Math.max(0, (int) (Math.floor(average) - _balancingConstraint.topicReplicaBalanceMinGap()));
```

Comment:
Shouldn't it be (swapping `maxLimit` and `minLimit`):
```
maxLimit = Math.max(0, (int) (Math.floor(average) - _balancingConstraint.topicReplicaBalanceMinGap()));
minLimit = Math.max(0, (int) (Math.floor(average) - _balancingConstraint.topicReplicaBalanceMaxGap()));
```
In the above way, `maxLimit` >= `minLimit`. Otherwise, it would be the other way (`maxLimit` <= `minLimit`), and the below return statement will always return `minLimit`

```
return Math.max(minLimit, Math.min(computedLimit, maxLimit));
```

Reply:
Absolutely -- great catch!